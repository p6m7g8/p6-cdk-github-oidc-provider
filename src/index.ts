export * from './p6-cdk-github-oidc-provider'
