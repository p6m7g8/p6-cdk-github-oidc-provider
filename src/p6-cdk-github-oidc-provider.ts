import { Stack } from 'aws-cdk-lib'
import * as iam from 'aws-cdk-lib/aws-iam'
import { Construct } from 'constructs'

export interface IP6CDKGithubOidcProviderProps {
  readonly repo: string
}

export class P6CDKGithubOidcProvider extends Construct {
  public readonly roleArn: string

  constructor(scope: Construct, id: string, props: IP6CDKGithubOidcProviderProps) {
    super(scope, id)

    const providerArn = `arn:aws:iam::${Stack.of(this).account}:oidc-provider/token.actions.githubusercontent.com`

    const existingOidcProvider = iam.OpenIdConnectProvider.fromOpenIdConnectProviderArn(this, 'ExistingOidcProvider', providerArn)

    // Create the OIDC Provider for GitHub
    const oidcProvider = existingOidcProvider ?? new iam.OpenIdConnectProvider(this, 'GithubOidc', {
      url: 'https://token.actions.githubusercontent.com',
      clientIds: ['sts.amazonaws.com'],
      thumbprints: ['ffffffffffffffffffffffffffffffffffffffff'],
    })

    // Create the IAM Role for GitHub OIDC Authentication
    const role = new iam.Role(this, 'GithubOidcRole', {
      assumedBy: new iam.FederatedPrincipal(
        oidcProvider.openIdConnectProviderArn,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            [`token.actions.githubusercontent.com:sub`]: `repo:${props.repo}:*`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    })

    // Expose the role ARN
    this.roleArn = role.roleArn
  }
}
