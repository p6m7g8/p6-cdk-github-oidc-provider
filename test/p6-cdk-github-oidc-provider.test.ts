import * as cdk from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { P6CDKGithubOidcProvider } from '../src'

// Test: Snapshot test for GithubOidcProvider Construct
it('snapshot test for GithubOidcProvider', () => {
  // Set up a CDK app and stack
  const app = new cdk.App()
  const stack = new cdk.Stack(app, 'TestStack')

  // Instantiate the GithubOidcProvider construct
  new P6CDKGithubOidcProvider(stack, 'GithubOidcTest', {
    repo: 'p6m7g8/p6m7g8.org',
  })

  // Generate the CloudFormation template for the stack
  const template = Template.fromStack(stack)

  // Expect the generated template to match the snapshot
  expect(template.toJSON()).toMatchSnapshot()
})
