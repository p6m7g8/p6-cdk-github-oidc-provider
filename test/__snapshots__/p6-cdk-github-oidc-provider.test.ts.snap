// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for GithubOidcProvider 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GithubOidcTestGithubOidcRoleB8226496": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:p6m7g8/p6m7g8.org:*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":oidc-provider/token.actions.githubusercontent.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
